File: src/plan_cascade/llm/providers/openai.py
"""
OpenAI LLM Provider

Implementation of LLMProvider for OpenAI's GPT models.
Uses the openai SDK for API communication.
"""

import asyncio
import os
from typing import Any, Dict, List, Optional, Union

from ..base import (
    LLMProvider,
    LLMResponse,
    ToolCall,
    TokenUsage,
    LLMError,
    RateLimitError,
    AuthenticationError,
    ModelNotFoundError,
)


class OpenAIProvider(LLMProvider):
    """
    LLM provider for OpenAI's GPT models.

    Supports:
    - GPT-4o, GPT-4 Turbo, GPT-4, GPT-3.5 Turbo
    - Function/tool calling
    - Automatic retry with exponential backoff

    Example:
        provider = OpenAIProvider(api_key="sk-...")
        response = await provider.complete([
            {"role": "user", "content": "Hello!"}
        ])
    """

    # Supported OpenAI models
    MODELS = [
        "gpt-4o",
        "gpt-4o-mini",
        "gpt-4-turbo",
        "gpt-4",
        "gpt-3.5-turbo",
        "o1-preview",
        "o1-mini",
    ]

    DEFAULT_MODEL = "gpt-4o"

    def __init__(
        self,
        api_key: Optional[str] = None,
        model: Optional[str] = None,
        base_url: Optional[str] = None,
        organization: Optional[str] = None,
        max_retries: int = 3,
        timeout: float = 120.0,
        **kwargs: Any
    ):
        """
        Initialize the OpenAI provider.

        Args:
            api_key: OpenAI API key (uses OPENAI_API_KEY env var if not provided)
            model: Model identifier
            base_url: Custom API base URL (for Azure or other compatible APIs)
            organization: OpenAI organization ID
            max_retries: Maximum number of retries
            timeout: Request timeout in seconds
            **kwargs: Additional configuration
        """
        api_key = api_key or os.environ.get("OPENAI_API_KEY")
        super().__init__(api_key=api_key, model=model, base_url=base_url, **kwargs)

        self.organization = organization or os.environ.get("OPENAI_ORG_ID")
        self.max_retries = max_retries
        self.timeout = timeout
        self._client = None

    def _get_client(self):
        """Get or create the OpenAI client."""
        if self._client is None:
            try:
                from openai import AsyncOpenAI
            except ImportError:
                raise LLMError(
                    "openai package not installed. Install with: pip install openai",
                    provider="openai"
                )

            client_kwargs = {}
            if self.api_key:
                client_kwargs["api_key"] = self.api_key
            if self.base_url:
                client_kwargs["base_url"] = self.base_url
            if self.organization:
                client_kwargs["organization"] = self.organization

            self._client = AsyncOpenAI(**client_kwargs)

        return self._client

    async def complete(
        self,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Dict[str, Any]]] = None,
        tool_choice: Optional[Union[str, Dict[str, Any]]] = None,
        temperature: float = 0.7,
        max_tokens: Optional[int] = None,
        **kwargs: Any
    ) -> LLMResponse:
        """
        Send a completion request to OpenAI.

        Args:
            messages: List of message dictionaries
            tools: Optional list of tool definitions
            tool_choice: Tool choice ("auto", "none", "required", or specific tool)
            temperature: Sampling temperature
            max_tokens: Maximum tokens to generate
            **kwargs: Additional parameters

        Returns:
            LLMResponse with the model's response
        """
        client = self._get_client()

        # Build request parameters
        params = {
            "model": self.model,
            "messages": self._format_messages(messages),
            "temperature": temperature,
        }

        if max_tokens:
            params["max_tokens"] = max_tokens

        # Add tools if provided
        if tools:
            params["tools"] = self._format_tools(tools)
            if tool_choice:
                params["tool_choice"] = self._format_tool_choice(tool_choice)

        # Execute with retry
        last_error = None
        for attempt in range(self.max_retries):
            try:
                response = await client.chat.completions.create(**params)
                return self._parse_response(response)
            except Exception as e:
                last_error = e
                error_msg = str(e).lower()

                # Check for rate limit
                if "rate" in error_msg and "limit" in error_msg:
                    wait_time = (2 ** attempt) * 1.0
                    await asyncio.sleep(wait_time)
                    continue

                # Check for authentication error
                if "auth" in error_msg or "key" in error_msg or "401" in str(e):
                    raise AuthenticationError(
                        f"OpenAI authentication failed: {e}",
                        provider="openai",
                        status_code=401
                    )

                # Check for model not found
                if "model" in error_msg and ("not found" in error_msg or "does not exist" in error_msg):
                    raise ModelNotFoundError(
                        f"Model not found: {self.model}",
                        provider="openai"
                    )

                # Other errors - retry if we have attempts left
                if attempt < self.max_retries - 1:
                    await asyncio.sleep(1.0)
                    continue

                raise LLMError(f"OpenAI API error: {e}", provider="openai")

        raise LLMError(f"Max retries exceeded: {last_error}", provider="openai")

    def _format_messages(self, messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Format messages for the OpenAI API."""
        formatted = []
        for msg in messages:
            role = msg.get("role", "user")
            content = msg.get("content", "")

            # Handle tool results
            if role == "tool":
                formatted.append({
                    "role": "tool",
                    "tool_call_id": msg.get("tool_call_id", ""),
                    "content": content,
                })
            elif role == "assistant" and msg.get("tool_calls"):
                # Assistant message with tool calls
                tool_calls = []
                for tc in msg.get("tool_calls", []):
                    tool_calls.append({
                        "id": tc.get("id", ""),
                        "type": "function",
                        "function": {
                            "name": tc.get("name", ""),
                            "arguments": (
                                tc.get("arguments", {})
                                if isinstance(tc.get("arguments"), str)
                                else self._dict_to_json(tc.get("arguments", {}))
                            ),
                        }
                    })
                formatted.append({
                    "role": "assistant",
                    "content": content or None,
                    "tool_calls": tool_calls,
                })
            else:
                formatted.append({"role": role, "content": content})

        return formatted

    def _dict_to_json(self, d: Dict[str, Any]) -> str:
        """Convert dictionary to JSON string."""
        import json
        return json.dumps(d)

    def _format_tools(self, tools: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Format tool definitions for the OpenAI API."""
        formatted = []
        for tool in tools:
            formatted.append({
                "type": "function",
                "function": {
                    "name": tool.get("name", ""),
                    "description": tool.get("description", ""),
                    "parameters": tool.get("parameters", {}),
                }
            })
        return formatted

    def _format_tool_choice(
        self,
        tool_choice: Union[str, Dict[str, Any]]
    ) -> Union[str, Dict[str, Any]]:
        """Format tool choice for the OpenAI API."""
        if isinstance(tool_choice, str):
            if tool_choice in ("auto", "none", "required"):
                return tool_choice
            else:
                # Specific function
                return {
                    "type": "function",
                    "function": {"name": tool_choice}
                }
        return tool_choice

    def _parse_response(self, response: Any) -> LLMResponse:
        """Parse the OpenAI API response."""
        choice = response.choices[0]
        message = choice.message

        content = message.content or ""
        tool_calls = []

        if message.tool_calls:
            import json
            for tc in message.tool_calls:
                try:
                    args = json.loads(tc.function.arguments)
                except json.JSONDecodeError:
                    args = {"raw": tc.function.arguments}

                tool_calls.append(ToolCall(
                    id=tc.id,
                    name=tc.function.name,
                    arguments=args,
                ))

        # Map finish reason
        stop_reason_map = {
            "stop": "end_turn",
            "tool_calls": "tool_use",
            "length": "max_tokens",
            "content_filter": "end_turn",
        }
        stop_reason = stop_reason_map.get(choice.finish_reason, "end_turn")

        # Parse usage
        usage = None
        if response.usage:
            usage = TokenUsage(
                input_tokens=response.usage.prompt_tokens,
                output_tokens=response.usage.completion_tokens,
                total_tokens=response.usage.total_tokens,
            )

        return LLMResponse(
            content=content,
            tool_calls=tool_calls,
            stop_reason=stop_reason,
            usage=usage,
            model=response.model,
        )

    def get_name(self) -> str:
        """Get the provider name."""
        return "openai"

    def get_default_model(self) -> str:
        """Get the default model."""
        return self.DEFAULT_MODEL

    def get_supported_models(self) -> List[str]:
        """Get supported models."""
        return self.MODELS.copy()

    def validate_config(self) -> bool:
        """Validate the provider configuration."""
        if not self.api_key:
            raise ValueError(
                "OpenAI API key is required. "
                "Set OPENAI_API_KEY environment variable or pass api_key parameter."
            )
        return True

File: desktop/src-tauri/crates/llm/src/openai.rs
//! OpenAI Provider
//!
//! Implementation of the LlmProvider trait for OpenAI's API.
//! Supports GPT-4, o1, and o3 models with tool calling and reasoning.

use async_trait::async_trait;
use futures_util::StreamExt;
use openai_api_rs::v1::chat_completion::chat_completion_stream::ChatCompletionStreamResponse;
use serde::Deserialize;
use tokio::sync::mpsc;

use super::openai_compat::{
    build_client, build_openai_compatible_messages, map_api_error, value_to_chat_request,
    value_to_chat_stream_request,
};
use super::provider::LlmProvider;
use super::types::{
    LlmError, LlmRequestOptions, LlmResponse, LlmResult, Message, ProviderConfig, StopReason,
    ToolCall, ToolCallMode, ToolDefinition, UsageStats,
};

#[cfg(test)]
use super::types::{MessageContent, MessageRole};
use plan_cascade_core::streaming::UnifiedStreamEvent;

/// Default OpenAI API endpoint
const OPENAI_API_URL: &str = "https://api.openai.com/v1/chat/completions";

/// OpenAI provider
pub struct OpenAIProvider {
    config: ProviderConfig,
}

impl OpenAIProvider {
    /// Create a new OpenAI provider with the given configuration
    pub fn new(config: ProviderConfig) -> Self {
        Self { config }
    }

    fn build_compat_client(&self) -> LlmResult<openai_api_rs::v1::api::OpenAIClient> {
        build_client(&self.config, "openai", OPENAI_API_URL, false)
    }

    /// Check if model supports reasoning (o1/o3 models)
    fn model_supports_reasoning(&self) -> bool {
        let model = self.config.model.to_lowercase();
        model.starts_with("o1") || model.starts_with("o3")
    }

    /// Build the request body for the API
    fn build_request_body(
        &self,
        messages: &[Message],
        system: Option<&str>,
        tools: &[ToolDefinition],
        stream: bool,
        request_options: &LlmRequestOptions,
    ) -> serde_json::Value {
        let mut body = serde_json::json!({
            "model": self.config.model,
            "max_tokens": self.config.max_tokens,
            "stream": stream,
        });

        // Add temperature (not for o1/o3 models)
        if !self.model_supports_reasoning() {
            body["temperature"] = serde_json::json!(request_options
                .temperature_override
                .unwrap_or(self.config.temperature));
        }

        // Add reasoning effort for o1/o3 models
        if self.model_supports_reasoning() {
            if let Some(effort) = request_options
                .reasoning_effort_override
                .as_ref()
                .or(self.config.reasoning_effort.as_ref())
            {
                body["reasoning_effort"] = serde_json::json!(effort);
            }
        }

        body["messages"] = serde_json::json!(build_openai_compatible_messages(messages, system));

        // Add tools if provided
        if !tools.is_empty() {
            let openai_tools: Vec<serde_json::Value> =
                tools.iter().map(|t| self.tool_to_openai(t)).collect();
            body["tools"] = serde_json::json!(openai_tools);
            if matches!(request_options.tool_call_mode, ToolCallMode::Required) {
                body["tool_choice"] = serde_json::json!("required");
            }
        }

        // Add stream options for usage in streaming
        if stream {
            body["stream_options"] = serde_json::json!({
                "include_usage": true
            });
        }

        body
    }

    #[cfg(test)]
    /// Convert a Message to OpenAI API format
    fn message_to_openai(&self, message: &Message) -> serde_json::Value {
        let role = match message.role {
            MessageRole::User => "user",
            MessageRole::Assistant => "assistant",
            MessageRole::System => "system",
        };

        // Check if message contains tool calls or tool results
        let has_tool_calls = message
            .content
            .iter()
            .any(|c| matches!(c, MessageContent::ToolUse { .. }));
        let has_tool_results = message.content.iter().any(|c| {
            matches!(
                c,
                MessageContent::ToolResult { .. } | MessageContent::ToolResultMultimodal { .. }
            )
        });

        if has_tool_results {
            // Tool results are sent as separate messages in OpenAI format
            let mut result_msg = serde_json::json!({
                "role": "tool"
            });

            for content in &message.content {
                match content {
                    MessageContent::ToolResult {
                        tool_use_id,
                        content,
                        ..
                    } => {
                        result_msg["tool_call_id"] = serde_json::json!(tool_use_id);
                        result_msg["content"] = serde_json::json!(content);
                        break;
                    }
                    MessageContent::ToolResultMultimodal {
                        tool_use_id,
                        content,
                        ..
                    } => {
                        // OpenAI tool results are text-only; extract text and include image as data URI
                        let mut parts = Vec::new();
                        for block in content {
                            match block {
                                super::types::ContentBlock::Text { text } => {
                                    parts.push(text.clone());
                                }
                                super::types::ContentBlock::Image { media_type, data } => {
                                    parts.push(format!(
                                        "[Image: data:{};base64,<{} bytes>]",
                                        media_type,
                                        data.len()
                                    ));
                                }
                            }
                        }
                        result_msg["tool_call_id"] = serde_json::json!(tool_use_id);
                        result_msg["content"] = serde_json::json!(parts.join("\n"));
                        break;
                    }
                    _ => {}
                }
            }

            return result_msg;
        }

        if has_tool_calls {
            let tool_calls: Vec<serde_json::Value> = message
                .content
                .iter()
                .filter_map(|c| {
                    if let MessageContent::ToolUse { id, name, input } = c {
                        Some(serde_json::json!({
                            "id": id,
                            "type": "function",
                            "function": {
                                "name": name,
                                "arguments": input.to_string()
                            }
                        }))
                    } else {
                        None
                    }
                })
                .collect();

            let text_content: String = message
                .content
                .iter()
                .filter_map(|c| {
                    if let MessageContent::Text { text } = c {
                        Some(text.as_str())
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>()
                .join("\n");

            let mut msg = serde_json::json!({
                "role": role,
                "tool_calls": tool_calls
            });

            // Always include content field — some OpenAI-compatible APIs
            // require it even when the assistant only emits tool calls.
            if text_content.is_empty() {
                msg["content"] = serde_json::Value::Null;
            } else {
                msg["content"] = serde_json::json!(text_content);
            }

            return msg;
        }

        // Check if message contains images (multimodal)
        let has_images = message
            .content
            .iter()
            .any(|c| matches!(c, MessageContent::Image { .. }));

        if has_images {
            // Build multimodal content array for OpenAI vision
            let content_parts: Vec<serde_json::Value> = message
                .content
                .iter()
                .filter_map(|c| match c {
                    MessageContent::Text { text } => Some(serde_json::json!({
                        "type": "text",
                        "text": text
                    })),
                    MessageContent::Image { media_type, data } => Some(serde_json::json!({
                        "type": "image_url",
                        "image_url": {
                            "url": format!("data:{};base64,{}", media_type, data)
                        }
                    })),
                    _ => None,
                })
                .collect();

            return serde_json::json!({
                "role": role,
                "content": content_parts
            });
        }

        // Simple text message
        let text_content: String = message
            .content
            .iter()
            .filter_map(|c| {
                if let MessageContent::Text { text } = c {
                    Some(text.as_str())
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
            .join("\n");

        serde_json::json!({
            "role": role,
            "content": text_content
        })
    }

    /// Convert a ToolDefinition to OpenAI API format
    fn tool_to_openai(&self, tool: &ToolDefinition) -> serde_json::Value {
        serde_json::json!({
            "type": "function",
            "function": {
                "name": tool.name,
                "description": tool.description,
                "parameters": tool.input_schema
            }
        })
    }

    /// Parse a response from OpenAI API
    fn parse_response(&self, response: &OpenAIResponse) -> LlmResponse {
        let choice = response.choices.first();

        let mut content = None;
        let mut thinking = None;
        let mut tool_calls = Vec::new();

        if let Some(choice) = choice {
            if let Some(msg) = &choice.message {
                content = msg.content.clone();
                thinking = msg.reasoning_content.clone();

                if let Some(tcs) = &msg.tool_calls {
                    for tc in tcs {
                        let arguments: serde_json::Value =
                            serde_json::from_str(&tc.function.arguments)
                                .unwrap_or(serde_json::Value::Null);

                        tool_calls.push(ToolCall {
                            id: tc.id.clone(),
                            name: tc.function.name.clone(),
                            arguments,
                        });
                    }
                }
            }
        }

        let stop_reason = choice
            .and_then(|c| c.finish_reason.as_ref())
            .map(|r| StopReason::from(r.as_str()))
            .unwrap_or(StopReason::EndTurn);

        let usage = response
            .usage
            .as_ref()
            .map(|u| UsageStats {
                input_tokens: u.prompt_tokens,
                output_tokens: u.completion_tokens,
                thinking_tokens: u.reasoning_tokens,
                cache_read_tokens: None,
                cache_creation_tokens: None,
            })
            .unwrap_or_default();

        LlmResponse {
            content,
            thinking,
            tool_calls,
            stop_reason,
            usage,
            model: response.model.clone(),
            search_citations: Vec::new(),
        }
    }
}

#[async_trait]
impl LlmProvider for OpenAIProvider {
    fn name(&self) -> &'static str {
        "openai"
    }

    fn model(&self) -> &str {
        &self.config.model
    }

    fn supports_thinking(&self) -> bool {
        self.model_supports_reasoning()
    }

    fn supports_tools(&self) -> bool {
        true
    }

    fn supports_multimodal(&self) -> bool {
        true
    }

    fn context_window(&self) -> u32 {
        let model = self.config.model.to_lowercase();
        if model.contains("o1") || model.contains("o3") || model.contains("o4") {
            200_000 // o1, o3-mini, o3, o4-mini: 200k
        } else if model.contains("gpt-4o")
            || model.contains("gpt-4-turbo")
            || model.contains("gpt-4.1")
        {
            128_000 // GPT-4o, GPT-4-turbo, GPT-4.1: 128k
        } else if model.contains("gpt-4-32k") {
            32_768
        } else if model.contains("gpt-4") {
            8_192 // GPT-4 (original): 8k
        } else if model.contains("gpt-3.5") {
            16_384 // GPT-3.5-turbo: 16k
        } else {
            128_000 // Default for newer/unknown models
        }
    }

    async fn send_message(
        &self,
        messages: Vec<Message>,
        system: Option<String>,
        tools: Vec<ToolDefinition>,
        request_options: LlmRequestOptions,
    ) -> LlmResult<LlmResponse> {
        let body = self.build_request_body(
            &messages,
            system.as_deref(),
            &tools,
            false,
            &request_options,
        );
        let request = value_to_chat_request("openai", body)?;
        let mut client = self.build_compat_client()?;

        let response = client
            .chat_completion(request)
            .await
            .map_err(|e| map_api_error("openai", e))?;

        let openai_response: OpenAIResponse =
            serde_json::from_value(serde_json::to_value(response).map_err(|e| {
                LlmError::ParseError {
                    message: format!("Failed to serialize response: {}", e),
                }
            })?)
            .map_err(|e| LlmError::ParseError {
                message: format!("Failed to parse response: {}", e),
            })?;

        Ok(self.parse_response(&openai_response))
    }

    async fn stream_message(
        &self,
        messages: Vec<Message>,
        system: Option<String>,
        tools: Vec<ToolDefinition>,
        tx: mpsc::Sender<UnifiedStreamEvent>,
        request_options: LlmRequestOptions,
    ) -> LlmResult<LlmResponse> {
        let body =
            self.build_request_body(&messages, system.as_deref(), &tools, true, &request_options);
        let request = value_to_chat_stream_request("openai", body)?;
        let mut client = self.build_compat_client()?;

        let mut stream = client
            .chat_completion_stream(request)
            .await
            .map_err(|e| map_api_error("openai", e))?;

        let mut accumulated_content = String::new();
        let mut accumulated_thinking = String::new();
        let mut tool_calls = Vec::new();
        let usage = UsageStats::default();
        let mut stop_reason = StopReason::EndTurn;
        let mut in_thinking = false;

        let mut pending_tools: std::collections::HashMap<String, (Option<String>, String)> =
            std::collections::HashMap::new();
        let mut pending_order: Vec<String> = Vec::new();

        while let Some(event) = stream.next().await {
            match event {
                ChatCompletionStreamResponse::Content(content) => {
                    if !content.is_empty() {
                        accumulated_content.push_str(&content);
                        let _ = tx.send(UnifiedStreamEvent::TextDelta { content }).await;
                    }
                }
                ChatCompletionStreamResponse::ReasoningContent(content) => {
                    if !content.is_empty() {
                        if !in_thinking {
                            in_thinking = true;
                            let _ = tx
                                .send(UnifiedStreamEvent::ThinkingStart { thinking_id: None })
                                .await;
                        }
                        accumulated_thinking.push_str(&content);
                        let _ = tx
                            .send(UnifiedStreamEvent::ThinkingDelta {
                                content,
                                thinking_id: None,
                            })
                            .await;
                    }
                }
                ChatCompletionStreamResponse::ToolCall(chunks) => {
                    for chunk in chunks {
                        let id = chunk.id;
                        if id.is_empty() {
                            continue;
                        }
                        if !pending_tools.contains_key(&id) {
                            pending_order.push(id.clone());
                        }
                        let entry = pending_tools
                            .entry(id)
                            .or_insert_with(|| (None, String::new()));
                        if let Some(name) = chunk.function.name {
                            if !name.is_empty() {
                                entry.0 = Some(name);
                            }
                        }
                        if let Some(arguments) = chunk.function.arguments {
                            entry.1.push_str(&arguments);
                        }
                    }
                }
                ChatCompletionStreamResponse::Done => break,
            }
        }

        if in_thinking {
            let _ = tx
                .send(UnifiedStreamEvent::ThinkingEnd { thinking_id: None })
                .await;
        }

        for id in pending_order {
            if let Some((name, args)) = pending_tools.remove(&id) {
                if let Some(name) = name {
                    if let Ok(arguments) = serde_json::from_str::<serde_json::Value>(&args) {
                        tool_calls.push(ToolCall {
                            id,
                            name,
                            arguments,
                        });
                    }
                }
            }
        }

        if !tool_calls.is_empty() {
            stop_reason = StopReason::ToolUse;
        }

        Ok(LlmResponse {
            content: if accumulated_content.is_empty() {
                None
            } else {
                Some(accumulated_content)
            },
            thinking: if accumulated_thinking.is_empty() {
                None
            } else {
                Some(accumulated_thinking)
            },
            tool_calls,
            stop_reason,
            usage,
            model: self.config.model.clone(),
            search_citations: Vec::new(),
        })
    }

    async fn health_check(&self) -> LlmResult<()> {
        let mut client = self.build_compat_client()?;
        client
            .list_models()
            .await
            .map(|_| ())
            .map_err(|e| map_api_error("openai", e))
    }

    fn config(&self) -> &ProviderConfig {
        &self.config
    }

    async fn list_models(&self) -> LlmResult<Option<Vec<String>>> {
        let mut client = self.build_compat_client()?;
        let response = client
            .list_models()
            .await
            .map_err(|e| map_api_error("openai", e))?;

        let models: Vec<String> = response
            .data
            .iter()
            .filter_map(|m| m.id.clone())
            .filter(|id| id.starts_with("gpt") || id.starts_with("o1") || id.starts_with("o3"))
            .collect();

        Ok(Some(models))
    }
}

/// OpenAI API response format
#[derive(Debug, Deserialize)]
struct OpenAIResponse {
    model: String,
    choices: Vec<Choice>,
    usage: Option<ResponseUsage>,
}

#[derive(Debug, Deserialize)]
struct Choice {
    message: Option<ResponseMessage>,
    finish_reason: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ResponseMessage {
    content: Option<String>,
    reasoning_content: Option<String>,
    tool_calls: Option<Vec<ResponseToolCall>>,
}

#[derive(Debug, Deserialize)]
struct ResponseToolCall {
    id: String,
    function: ResponseFunction,
}

#[derive(Debug, Deserialize)]
struct ResponseFunction {
    name: String,
    arguments: String,
}

#[derive(Debug, Deserialize)]
struct ResponseUsage {
    prompt_tokens: u32,
    completion_tokens: u32,
    #[serde(default)]
    reasoning_tokens: Option<u32>,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_config() -> ProviderConfig {
        ProviderConfig {
            provider: super::super::types::ProviderType::OpenAI,
            api_key: Some("sk-test".to_string()),
            model: "gpt-4".to_string(),
            ..Default::default()
        }
    }

    #[test]
    fn test_provider_creation() {
        let provider = OpenAIProvider::new(test_config());
        assert_eq!(provider.name(), "openai");
        assert_eq!(provider.model(), "gpt-4");
        assert!(!provider.supports_thinking()); // GPT-4 doesn't support reasoning
        assert!(provider.supports_tools());
    }

    #[test]
    fn test_o1_supports_reasoning() {
        let config = ProviderConfig {
            model: "o1-preview".to_string(),
            ..test_config()
        };
        let provider = OpenAIProvider::new(config);
        assert!(provider.supports_thinking());
    }

    #[test]
    fn test_message_conversion() {
        let provider = OpenAIProvider::new(test_config());
        let message = Message::user("Hello!");

        let openai_msg = provider.message_to_openai(&message);
        assert_eq!(openai_msg["role"], "user");
        assert_eq!(openai_msg["content"], "Hello!");
    }

    #[test]
    fn test_tool_conversion() {
        let provider = OpenAIProvider::new(test_config());
        let tool = ToolDefinition {
            name: "get_weather".to_string(),
            description: "Get weather".to_string(),
            input_schema: super::super::types::ParameterSchema::object(
                None,
                std::collections::HashMap::new(),
                vec![],
            ),
        };

        let openai_tool = provider.tool_to_openai(&tool);
        assert_eq!(openai_tool["type"], "function");
        assert_eq!(openai_tool["function"]["name"], "get_weather");
    }
}
